data/transactions/verify/txn.go

```diff
@@ -448,3 +456,397 @@ func (w *worksetBuilder) next() (txnGroups [][]transactions.SignedTxn) {
+}
+
+// OnNewBlock implements the interface to subscribe to new block notifications from the ledger
+func (nbw *NewBlockWatcher) OnNewBlock(block bookkeeping.Block, delta ledgercore.StateDelta) {
+	if nbw.blkHeader.Round >= block.BlockHeader.Round {
+		return
+	}
+	nbw.mu.Lock()
+	defer nbw.mu.Unlock()
+	nbw.blkHeader = block.BlockHeader
+}
+
+func (nbw *NewBlockWatcher) getBlockHeader() (bh bookkeeping.BlockHeader) {
+	nbw.mu.RLock()
+	defer nbw.mu.RUnlock()
+	return nbw.blkHeader
+}
+
+type unverifiedElementList struct {
+	elementList []UnverifiedElement
+	nbw         *NewBlockWatcher
+	ledger      logic.LedgerForSignature
+}
+
+func makeUnverifiedElementList(nbw *NewBlockWatcher, ledger logic.LedgerForSignature) (uel unverifiedElementList) {
+	uel.nbw = nbw
+	uel.ledger = ledger
+	uel.elementList = make([]UnverifiedElement, 0)
+	return
+}
+
+func makeSingleUnverifiedElement(nbw *NewBlockWatcher, ledger logic.LedgerForSignature, ue UnverifiedElement) (uel unverifiedElementList) {
+	uel.nbw = nbw
+	uel.ledger = ledger
+	uel.elementList = []UnverifiedElement{ue}
+	return
+}
+
+type batchLoad struct {
+	txnGroups      [][]transactions.SignedTxn
+	groupCtxs      []*GroupContext
+	elementContext []interface{}
+	messagesForTxn []int
+}
+
+func makeBatchLoad() (bl batchLoad) {
+	bl.txnGroups = make([][]transactions.SignedTxn, 0)
+	bl.groupCtxs = make([]*GroupContext, 0)
+	bl.elementContext = make([]interface{}, 0)
+	bl.messagesForTxn = make([]int, 0)
+	return bl
+}
+
+func (bl *batchLoad) addLoad(txngrp []transactions.SignedTxn, gctx *GroupContext, eltctx interface{}, numBatchableSigs int) {
+	bl.txnGroups = append(bl.txnGroups, txngrp)
+	bl.groupCtxs = append(bl.groupCtxs, gctx)
+	bl.elementContext = append(bl.elementContext, eltctx)
+	bl.messagesForTxn = append(bl.messagesForTxn, numBatchableSigs)
+
+}
+
+// waitForNextTxnDuration is the time to wait before sending the batch for evaluation.
+// this does not mean that every batch will wait this long. when the load level is high,
+// the batch will fill up quickly and will be sent for evaluation before this timeout
+const waitForNextTxnDuration = 1 * time.Millisecond
+
+// waitForFirstTxnDuration is the time to wait for the first transaction in the batch
+var waitForFirstTxnDuration = 2000 * time.Millisecond
+
+// MakeStreamVerifier creates a new stream verifier and returns the chans used to send txn groups
+// to it and obtain the txn signature verification result from
+func MakeStreamVerifier(ctx context.Context, stxnChan <-chan UnverifiedElement, resultChan chan<- VerificationResult,
+	ledger logic.LedgerForSignature, nbw *NewBlockWatcher, verificationPool execpool.BacklogPool,
+	cache VerifiedTransactionCache) (sv *StreamVerifier) {
+
+	// limit the number of tasks queued to the execution pool
+	// the purpose of this parameter is to create bigger batches
+	// instead of having many smaller batching waiting in the execution pool
+	numberOfExecPoolSeats := verificationPool.GetParallelism() * 2
```

Comment:
> If we only have N workers, why do we want to limit the max enqueued batch tasks to N*2?

It is the convention in the code to take 2x, as you can see elsewhere 2x or 3x is used. Not a big deal here, will make more sense below. 

> Also, why do we need seats at all if we can control how big the batches using the max batch size and the batch flush timeout you've already implemented?

The purpose of the number of seats is to guess when the exec pool is saturated, and when it is saturated, not to send more tasks there. 
If the exec pool is saturated and a task is sent to it, the task will sit in the buffer and wait. Batch verifier can make a good use of this waiting time by keeping the batch in the loop, and adding more transactions to it. 
I.e. instead of having multiple small batches sitting in the exec pool buffer, they can be combined and sent as a single bigger batch to the exec pool for verification, instead of overloading it will small tasks.